File: C/Partie5/Exo23b.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <semaphore.h>
#include <unistd.h>
#include <time.h>

int N;         // nombre de threads (passé en argument)
int count = 0; // compteur d'arrivée
pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
sem_t barriere; // sémaphore pour bloquer les threads

void *thread_rdv(void *arg)
{
    int id = *(int *)arg;
    free(arg);
    for (int iter = 1; iter <= 5; iter++)
    {
        int duree = rand() % 5 + 1; // entre 1 et 5 secondes
        printf("Thread %d : début itération %d, durée %d s\n", id, iter, duree);
        sleep(duree);
        printf("Thread %d : fin itération %d, attente des autres\n", id, iter);

        // Barrière
        pthread_mutex_lock(&mutex);
        count++;
        if (count == N)
        {
            // dernier thread : libère les N-1 autres
            for (int i = 0; i < N - 1; i++)
                sem_post(&barriere);
            count = 0;
            pthread_mutex_unlock(&mutex);
        }
        else
        {
            pthread_mutex_unlock(&mutex);
            sem_wait(&barriere);
        }
    }
    printf("Thread %d terminé\n", id);
    return NULL;
}

int main(int argc, char *argv[])
{
    if (argc != 2)
    {
        fprintf(stderr, "Usage: %s N\n", argv[0]);
        exit(1);
    }
    N = atoi(argv[1]);
    if (N <= 0)
        exit(1);

    srand(time(NULL));
    sem_init(&barriere, 0, 0);

    pthread_t *threads = malloc(N * sizeof(pthread_t));
    for (int i = 0; i < N; i++)
    {
        int *id = malloc(sizeof(int));
        *id = i + 1;
        pthread_create(&threads[i], NULL, thread_rdv, id);
    }
    for (int i = 0; i < N; i++)
        pthread_join(threads[i], NULL);

    free(threads);
    sem_destroy(&barriere);
    pthread_mutex_destroy(&mutex);
    return 0;
}

File: Python/Partie5/Exo23b.py
import threading
import time
import random
import sys

def barriere(N, count, mutex, sem):
    mutex.acquire()
    count[0] += 1
    if count[0] == N:
        count[0] = 0
        for _ in range(N-1):
            sem.release()
        mutex.release()
    else:
        mutex.release()
        sem.acquire()

def thread_rdv(id, N, count, mutex, sem):
    for _ in range(5):
        duree = random.randint(1, 5)
        print(f"Thread {id} : début, durée {duree}s")
        time.sleep(duree)
        print(f"Thread {id} : arrivée")
        barriere(N, count, mutex, sem)
    print(f"Thread {id} terminé")

if len(sys.argv) != 2:
    print("Usage: python prog.py N")
    sys.exit(1)
N = int(sys.argv[1])
count = [0]
mutex = threading.Lock()
sem = threading.Semaphore(0)
threads = []
for i in range(N):
    t = threading.Thread(target=thread_rdv, args=(i+1, N, count, mutex, sem))
    t.start()
    threads.append(t)
for t in threads:
    t.join()